### ISSUE TITLE
Split Functions/Methods Definitions and Calls

### ISSUE THREAD

**Issue by lfenzo** [nvim-treesitter/nvim-treesitter#3210]
### Describe the highlighting problem

It would be good to have method/functions definition and calls in separate highlighting groups e.g. `TSMethodDefinition`, `TSFunctionDefinition`, `TSMethodCall` and `TSFunctionCall`, as it can be used to better emphasize the signature of methods and functions in their definitions and better distinguish them from their calls. This is something I noticed while porting my current colorscheme (which previously had this distinctions) to Tree-Sitter once the function definitions were often "visually diluted" in the middle of several other functions calls in each function block. In this context, I would like to make this definitions stand out a little more in the code blocks.

### Example snippet that causes the problem

Just as an example, consider the following code highlighted with the same colors for and functions and the method definition:

```
void runOnFunction() override {
    llvm::SmallVector<ParallelOp, 2> innermostPloops;
    mlir::getInnermostParallelLoops(this->getFunction().getOperation(),
                                    innermostPloops);
    for (ParallelOp ploop : innermostPloops) {
      // of the arguments is a constant), i.e. it's not inside `shape.assuming`.
      if (ploop->getParentOfType<mlir::shape::AssumingOp>() != nullptr) {
        tileParallelLoop(ploop, tile_sizes_);
        continue;
      }
      auto tiled_loops = tileParallelLoop(ploop, outer_tile_);
      // Tile twice if the inner_tile is non-empty.
      if (!inner_tile_.empty()) {
        tileParallelLoop(tiled_loops.second, inner_tile_);
      }
    }
  }
```

### Tree-sitter parsing result

For the snippet above, this is the output of `TSPlaygroundToggle`:

```
function_definition [0, 0] - [16, 3]
  type: primitive_type [0, 0] - [0, 4]
  declarator: function_declarator [0, 5] - [0, 29]
    declarator: identifier [0, 5] - [0, 18]
    parameters: parameter_list [0, 18] - [0, 20]
    virtual_specifier [0, 21] - [0, 29]
  body: compound_statement [0, 30] - [16, 3]
    declaration [1, 4] - [1, 53]
      type: qualified_identifier [1, 4] - [1, 36]
        scope: namespace_identifier [1, 4] - [1, 8]
        name: template_type [1, 10] - [1, 36]
          name: type_identifier [1, 10] - [1, 21]
          arguments: template_argument_list [1, 21] - [1, 36]
            type_descriptor [1, 22] - [1, 32]
              type: type_identifier [1, 22] - [1, 32]
            number_literal [1, 34] - [1, 35]
      declarator: identifier [1, 37] - [1, 52]
    expression_statement [2, 4] - [3, 53]
      call_expression [2, 4] - [3, 52]
        function: qualified_identifier [2, 4] - [2, 35]
          scope: namespace_identifier [2, 4] - [2, 8]
          name: identifier [2, 10] - [2, 35]
        arguments: argument_list [2, 35] - [3, 52]
          call_expression [2, 36] - [2, 70]
            function: field_expression [2, 36] - [2, 68]
              argument: call_expression [2, 36] - [2, 55]
                function: field_expression [2, 36] - [2, 53]
                  argument: this [2, 36] - [2, 40]
                  field: field_identifier [2, 42] - [2, 53]
                arguments: argument_list [2, 53] - [2, 55]
              field: field_identifier [2, 56] - [2, 68]
            arguments: argument_list [2, 68] - [2, 70]
          identifier [3, 36] - [3, 51]
    for_range_loop [4, 4] - [15, 5]
      type: type_identifier [4, 9] - [4, 19]
      declarator: identifier [4, 20] - [4, 25]
      right: identifier [4, 28] - [4, 43]
      body: compound_statement [4, 45] - [15, 5]
        comment [5, 6] - [5, 80]
        if_statement [6, 6] - [9, 7]
          condition: condition_clause [6, 9] - [6, 71]
            value: binary_expression [6, 10] - [6, 70]
              left: call_expression [6, 10] - [6, 59]
                function: field_expression [6, 10] - [6, 57]
                  argument: identifier [6, 10] - [6, 15]
                  field: template_method [6, 17] - [6, 57]
                    name: field_identifier [6, 17] - [6, 32]
                    arguments: template_argument_list [6, 32] - [6, 57]
                      type_descriptor [6, 33] - [6, 56]
                        type: qualified_identifier [6, 33] - [6, 56]
                          scope: namespace_identifier [6, 33] - [6, 37]
                          name: qualified_identifier [6, 39] - [6, 56]
                            scope: namespace_identifier [6, 39] - [6, 44]
                            name: type_identifier [6, 46] - [6, 56]
                arguments: argument_list [6, 57] - [6, 59]
              right: nullptr [6, 63] - [6, 70]
          consequence: compound_statement [6, 72] - [9, 7]
            expression_statement [7, 8] - [7, 45]
              call_expression [7, 8] - [7, 44]
                function: identifier [7, 8] - [7, 24]
                arguments: argument_list [7, 24] - [7, 44]
                  identifier [7, 25] - [7, 30]
                  identifier [7, 32] - [7, 43]
            continue_statement [8, 8] - [8, 17]
        declaration [10, 6] - [10, 62]
          type: placeholder_type_specifier [10, 6] - [10, 10]
            auto [10, 6] - [10, 10]
          declarator: init_declarator [10, 11] - [10, 61]
            declarator: identifier [10, 11] - [10, 22]
            value: call_expression [10, 25] - [10, 61]
              function: identifier [10, 25] - [10, 41]
              arguments: argument_list [10, 41] - [10, 61]
                identifier [10, 42] - [10, 47]
                identifier [10, 49] - [10, 60]
        comment [11, 6] - [11, 51]
        if_statement [12, 6] - [14, 7]
          condition: condition_clause [12, 9] - [12, 31]
            value: unary_expression [12, 10] - [12, 30]
              argument: call_expression [12, 11] - [12, 30]
                function: field_expression [12, 11] - [12, 28]
                  argument: identifier [12, 11] - [12, 22]
                  field: field_identifier [12, 23] - [12, 28]
                arguments: argument_list [12, 28] - [12, 30]
          consequence: compound_statement [12, 32] - [14, 7]
            expression_statement [13, 8] - [13, 58]
              call_expression [13, 8] - [13, 57]
                function: identifier [13, 8] - [13, 24]
                arguments: argument_list [13, 24] - [13, 57]
                  field_expression [13, 25] - [13, 43]
                    argument: identifier [13, 25] - [13, 36]
                    field: field_identifier [13, 37] - [13, 43]
                  identifier [13, 45] - [13, 56]
```

### Example screenshot

Current Highlighting: 
![image](https://user-images.githubusercontent.com/49812445/180335508-dd13aa40-a938-4c72-85f4-2c654378df0c.png)

The colorscheme is [Tokyo Night](https://github.com/folke/tokyonight.nvim)

### Expected behavior

Having functions/methods definitions highlighted differently then their respective calls in the code. As suggested above, new groups, e.g. `TSMethodDefinition`, `TSFunctionDefinition`, `TSMethodCall` and `TSFunctionCall, to address this highlights would be extremely helpful.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - WARNING: `tree-sitter` executable not found (parser generator, only needed for :TSInstallFromGrammar, not required for :TSInstall)
  - OK: `node` found v16.15.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (GCC) 12.1.1 20220507 (Red Hat 12.1.1-1)
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - markdown       ✓ . ✓ . ✓ 
  - latex          ✓ . ✓ . ✓ 
  - python         ✓ ✓ ✓ ✓ ✓ 
  - lua            ✓ ✓ ✓ ✓ ✓ 
  - cpp            ✓ ✓ ✓ ✓ ✓ 
  - javascript     ✓ ✓ ✓ ✓ ✓ 
  - julia          ✓ ✓ ✓ ✓ ✓ 
  - go             ✓ ✓ ✓ ✓ ✓ 
  - java           ✓ ✓ . ✓ ✓ 
  - rust           ✓ ✓ ✓ ✓ ✓ 
  - c              ✓ ✓ ✓ ✓ ✓ 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
VIM v0.7.2
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -Wp,-U_FORTIFY_SOURCE -Wp,-D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -DNVIM_TS_HAS_SET_ALLOCATOR -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wdouble-promotion -Wmissing-noreturn -Wmissing-format-attribute -Wmissing-prototypes -Wimplicit-fallthrough -Wsuggest-attribute=pure -Wsuggest-attribute=const -Wsuggest-attribute=malloc -Wsuggest-attribute=cold -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/builddir/build/BUILD/neovim-0.7.2/redhat-linux-build/config -I/builddir/build/BUILD/neovim-0.7.2/src -I/usr/include -I/usr/include/luajit-2.1 -I/builddir/build/BUILD/neovim-0.7.2/redhat-linux-build/src/nvim/auto -I/builddir/build/BUILD/neovim-0.7.2/redhat-linux-build/include
Compiled by mockbuild@koji

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

**Comment by kyazdani42** [comment]
A method is a function in an object, to me they have the same semantic. I don't think adding new highlight groups would be an issue, but since we try to keep this repository as simple as we can, i think this could be added on your end. You'd have to augment the `vim.treesitter.highlighter.hl_map`:
```lua
vim.treesitter.highlight.hl_map['method.call'] = 'TSMethodCall'
```
then if your query
```scheme
(node) @method.call
```

**Comment by theHamsta** [comment]
I would be open to have those highlight groups, but we should try to add them to all languages. In fact, we even had `@function.call` but removed it in the beginning phase of the project to minimize the number of highlight groups. @lfenzo do you want to create a PR adding the captures for `@method.call`/`@function.call` to some top-tier languages when general feedback is positive?

**Comment by lfenzo** [comment]
> @lfenzo do you want to create a PR adding the captures for @method.call/@function.call to some top-tier languages when general feedback is positive?

I would like to, but honestly, I have no clue on how to do it. I've skimmed through the `queries/*/highlights.scm` files to get an idea of how they should look like (some of them even seem to already support the distinction between definition and call, like `queries/python/highlights.scm`) but that didn't help much. If anyone could provide some example of how to make these changes properly or point out resources or docs, that would be very helpful.

Some of the questions I got after looking into the highlight scheme files in `queries/` were:

1. Is there a list of definitions for the `@<ast-object>`s, like `@function`, `@constant`, `@constant.builtin`? Or can I just define them inside the `highlights.scm` and expect them to be recognized somewhere else in order to match the captures? For instance, in the python `highlights.scm`, it seems to me that some of the `@function`s could be simply replaced with `@function.call`, but I don't know if `@function.call` *exists* and can be recognized by highlight groups.

1. How do I create queries (and test them)? In the example `(node) @method.call` I'm assuming that `(node)` is something like the nodes appearing in the output of Tree-Sitter Playground, is this correct? 

1. In the case of a PR where can I link the objects `@function.call` with `TSFunctionCall` highlight group as mentioned by @kyazdani42?

I'm new to Tree-Sitter and up until now I have only used it to make colorschemes, but if anyone can help me with these changes I can create the PR addressing the `@function.call` and `@method.call`.

**Comment by clason** [comment]
You can see the list of currently allowed captures here: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights

But I think there's a slight misunderstanding: nobody expects you to write _new_ queries for many different languages; just to change the already existing queries to change, e.g., `@function` to `@function.call` where it makes sense.

**Comment by lfenzo** [comment]
Thanks @clason!

I will make these changes and soon submit a draft PR.

**Comment by lfenzo** [comment]
Since #3222 has been merged and #3262 created to track the languages with currently lack the changes in their syntax queries in order to captures function/method calls, I will take the liberty of closing this issue.
